src/core/utils.cpp

```diff
@@ -4,7 +4,7 @@
 char *strcat_alloc(char *left, const char *right) {
   if (!left) {
     char *res = (char *)Utils::malloc(strlen(right) + 1);
-    strncpy(res, right, strlen(right) + 1);
+    strncpy(res, right, strlen(res));
```

Comment:
That looks incorrect to me, as `res` is not initialized, it needs not be null-terminated. Please fix.

Reply:
Ideally one would replace the `log` in P3M with a `std::string` or `std::stringstream` and remove this function altogether.